Source/NSSpeechRecognizer.m

```diff
@@ -23,8 +23,163 @@
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSArray.h>
+#import <Foundation/NSThread.h>
+#import <Foundation/NSError.h>
+#import <Foundation/NSConnection.h>
+#import <Foundation/NSDistributedNotificationCenter.h>
 
-@implementation NSSpeechRecognizer
+#import "AppKit/NSWorkspace.h"
+
+id _speechRecognitionServer = nil;
+Class _speechRecognitionClass = nil;
+BOOL _serverLaunchTested = NO;
+
+#define SPEECH_RECOGNITION_SERVER @"GSSpeechRecognitionServer"
 
+@interface NSObject (GSSpeechRecognitionServer)
+- (NSSpeechRecognizer *)newRecognizer;
 @end
 
+@implementation NSSpeechRecognizer
+
++ (void) initialize
+{
+  _speechRecognitionClass = [NSSpeechRecognizer class];
+  _speechRecognitionServer = [[NSConnection rootProxyForConnectionWithRegisteredName: SPEECH_RECOGNITION_SERVER
+                                                                                host: nil] retain];
+  if (nil == _speechRecognitionServer)
+    {
+      NSWorkspace *ws = [NSWorkspace sharedWorkspace];
+      [ws launchApplication: SPEECH_RECOGNITION_SERVER
+                   showIcon: NO
+                 autolaunch: NO];
+    }
+}
+
+- (void) processNotification: (NSNotification *)note
+{
+  NSString *word = (NSString *)[note object];
+  NSEnumerator *en = [_commands objectEnumerator];
+  id obj = nil;
+  while((obj = [en nextObject]) != nil)
+    {
+      if ([[obj lowercaseString] isEqualToString: [word lowercaseString]])
+        {
+          [_delegate speechRecognizer: self
+                  didRecognizeCommand: word];
+        }
+    }
+}
+
+// Initialize
+- (instancetype) init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      [[NSDistributedNotificationCenter defaultCenter]
+        addObserver: self
+           selector: @selector(processNotification:)
+               name: GSSpeechRecognizerDidRecognizeWordNotification
+             object: nil];
+    }
+  return self;
+}
+
++ (id) allocWithZone: (NSZone *)aZone
+{
+  if (self == _speechRecognitionClass)
+    {
+      if (nil == _speechRecognitionServer && !_serverLaunchTested)
+        {
+          unsigned int i=0;
+          // Wait for up to five seconds  for the server to launch, then give up.
+          for (i=0 ; i<50 ; i++)
+            {
+              _speechRecognitionServer = [[NSConnection rootProxyForConnectionWithRegisteredName: SPEECH_RECOGNITION_SERVER
+                                                                                            host: nil] retain];
+              if (nil != _speechRecognitionServer)
+                {
+                  break;
+                }
+              [NSThread sleepForTimeInterval: 0.1];
+            }
+          
+          // Set a flag so we don't bother waiting for the speech recognition server to
+          // launch the next time if it didn't work this time.
+          _serverLaunchTested = YES;
+        }
+      // If there is no server, this will return nil
+      return [_speechRecognitionServer newRecognizer];
```

Comment:
I must admit I don't understand the logic here. If this method gets called on the NSSpeechRecognizer class we return an NSProxy to a server object. Doesn't this make all our local methods obsolete?